File: togo/internal/compiler/narrowing.go
package compiler

// Port of src/compiler/narrowing.ts: flow-aware nullness narrowing (nikeee/cappu#25).
// A syntactic dominator walk - not a full control-flow graph - that figures out what
// the code has proven about a local/parameter at a use site, the analog of
// TypeScript's getTypeOfSymbolAtLocation. Covers if/else guards, early-exit
// (if (x==null) return), && / || short-circuit, ternary, instanceof,
// Objects.requireNonNull / assert, and reassignment. Loops' back-edges and
// branch-merge of reassignments are not modeled.

// resolveRef resolves a bare reference node to its symbol (the checker's ResolveName).
type resolveRef func(node *Node) *Symbol

// exprNullnessFn is the provable nullness of a value expression (null literal, `new`,
// @NonNull, ...).
type exprNullnessFn func(node *Node) Nullness

// facts is what a boolean condition proves about a symbol when true / when false.
type facts struct {
	whenTrue  Nullness
	whenFalse Nullness
}

func isNullLiteral(n *Node) bool { return n.Kind == NullKeyword }

// clauseMatchesNull reports whether a switch clause has a `case null` label.
func clauseMatchesNull(clause *SwitchClauseData) bool {
	if clause.Labels == nil {
		return false
	}
	for _, l := range clause.Labels.Nodes {
		if l.Kind == NullKeyword {
			return true
		}
	}
	return false
}

func switchHasNullCase(clauses *NodeArray) bool {
	if clauses == nil {
		return false
	}
	for _, c := range clauses.Nodes {
		if clauseMatchesNull(c.AsSwitchClause()) {
			return true
		}
	}
	return false
}

// refersToSymbol reports whether node references symbol: a bare `x` (local/param/
// implicit-this field) or a `this.x` member access (a final field). `this`-only by
// design - `obj.x` for an arbitrary receiver shares the field symbol across receivers
// and must not narrow.
func refersToSymbol(node *Node, symbol *Symbol, resolve resolveRef) bool {
	if node.Kind == Identifier {
		return resolve(node) == symbol
	}
	if node.Kind == PropertyAccessExpression {
		return node.AsPropertyAccessExpression().Expression.Kind == ThisExpression && resolve(node) == symbol
	}
	return false
}

// calleeName is the simple (unqualified) name of a call's callee.
func calleeName(call *CallExpressionData) string {
	switch call.Expression.Kind {
	case Identifier:
		return call.Expression.AsIdentifier().Text
	case PropertyAccessExpression:
		return call.Expression.AsPropertyAccessExpression().Name.AsIdentifier().Text
	}
	return ""
}

func conditionImplies(cond *Node, symbol *Symbol, resolve resolveRef) facts {
	switch cond.Kind {
	case ParenthesizedExpression:
		return conditionImplies(cond.AsParenthesizedExpression().Expression, symbol, resolve)
	case PrefixUnaryExpression:
		u := cond.AsPrefixUnaryExpression()
		if u.Operator != ExclamationToken {
			return facts{}
		}
		f := conditionImplies(u.Operand, symbol, resolve)
		return facts{whenTrue: f.whenFalse, whenFalse: f.whenTrue}
	case BinaryExpression:
		b := cond.AsBinaryExpression()
		switch b.OperatorToken {
		case EqualsEqualsToken, ExclamationEqualsToken:
			isNullCheck := (refersToSymbol(b.Left, symbol, resolve) && isNullLiteral(b.Right)) ||
				(refersToSymbol(b.Right, symbol, resolve) && isNullLiteral(b.Left))
			if !isNullCheck {
				return facts{}
			}
			// `x == null`: true => null, false => non-null. `x != null`: the inverse.
			if b.OperatorToken == EqualsEqualsToken {
				return facts{whenTrue: NullnessNullable, whenFalse: NullnessNonNull}
			}
			return facts{whenTrue: NullnessNonNull, whenFalse: NullnessNullable}
		case AmpersandAmpersandToken:
			l := conditionImplies(b.Left, symbol, resolve)
			r := conditionImplies(b.Right, symbol, resolve)
			return facts{whenTrue: orNullness(l.whenTrue, r.whenTrue)}
		case BarBarToken:
			l := conditionImplies(b.Left, symbol, resolve)
			r := conditionImplies(b.Right, symbol, resolve)
			return facts{whenFalse: orNullness(l.whenFalse, r.whenFalse)}
		}
		return facts{}
	case InstanceofExpression:
		// `x instanceof T` is false for null, so true proves x non-null.
		if refersToSymbol(cond.AsInstanceofExpression().Expression, symbol, resolve) {
			return facts{whenTrue: NullnessNonNull}
		}
		return facts{}
	case CallExpression:
		call := cond.AsCallExpression()
		arg0 := firstArg(call)
		if arg0 == nil || !refersToSymbol(arg0, symbol, resolve) {
			return facts{}
		}
		switch calleeName(call) {
		case "nonNull": // Objects.nonNull(x)
			return facts{whenTrue: NullnessNonNull}
		case "isNull": // Objects.isNull(x)
			return facts{whenTrue: NullnessNullable, whenFalse: NullnessNonNull}
		}
		return facts{}
	}
	return facts{}
}

func orNullness(a, b Nullness) Nullness {
	if a != "" {
		return a
	}
	return b
}

func firstArg(call *CallExpressionData) *Node {
	if call.Arguments == nil || len(call.Arguments.Nodes) == 0 {
		return nil
	}
	return call.Arguments.Nodes[0]
}

// definitelyExits reports whether a statement always completes abruptly (so code
// after it is unreachable): return/throw/break/continue, or a block ending in one.
func definitelyExits(stmt *Node) bool {
	if stmt == nil {
		return false
	}
	switch stmt.Kind {
	case ReturnStatement, ThrowStatement, BreakStatement, ContinueStatement:
		return true
	case Block:
		s := stmt.AsBlock().Statements
		return s != nil && len(s.Nodes) > 0 && definitelyExits(s.Nodes[len(s.Nodes)-1])
	}
	return false
}

// stmtFactKind classifies a preceding statement's effect on a symbol.
type stmtFactKind int

const (
	factNone   stmtFactKind = iota // not relevant; keep looking
	factNarrow                     // proves a nullness
	factReset                      // an assignment whose value is unprovable; stop
)

type stmtFact struct {
	kind     stmtFactKind
	nullness Nullness
}

// scanPrecedingFacts scans stmts[0:before) backward for the nearest decisive fact
// about symbol (an assignment or guard); factNone when none touches it.
func scanPrecedingFacts(stmts []*Node, before int, symbol *Symbol, resolve resolveRef, exprNullness exprNullnessFn) stmtFact {
	for i := before - 1; i >= 0; i-- {
		if fact := precedingStatementFact(stmts[i], symbol, resolve, exprNullness); fact.kind != factNone {
			return fact
		}
	}
	return stmtFact{kind: factNone}
}

func precedingStatementFact(stmt *Node, symbol *Symbol, resolve resolveRef, exprNullness exprNullnessFn) stmtFact {
	switch stmt.Kind {
	case ExpressionStatement:
		expr := stmt.AsExpressionStatement().Expression
		switch expr.Kind {
		case AssignmentExpression:
			a := expr.AsAssignmentExpression()
			if a.OperatorToken == EqualsToken && refersToSymbol(a.Left, symbol, resolve) {
				if n := exprNullness(a.Right); n != "" {
					return stmtFact{kind: factNarrow, nullness: n}
				}
				return stmtFact{kind: factReset}
			}
		case CallExpression:
			call := expr.AsCallExpression()
			arg0 := firstArg(call)
			if calleeName(call) == "requireNonNull" && arg0 != nil && refersToSymbol(arg0, symbol, resolve) {
				return stmtFact{kind: factNarrow, nullness: NullnessNonNull}
			}
		}
	case AssertStatement:
		if conditionImplies(stmt.AsAssertStatement().Condition, symbol, resolve).whenTrue == NullnessNonNull {
			return stmtFact{kind: factNarrow, nullness: NullnessNonNull}
		}
	case IfStatement:
		ifs := stmt.AsIfStatement()
		facts := conditionImplies(ifs.Condition, symbol, resolve)
		// Early-exit: if (COND) <abrupt>;  -> after the if, only the fall-through
		// path remains (e.g. if (x==null) return; -> non-null).
		if ifs.ElseStatement == nil && definitelyExits(ifs.ThenStatement) && facts.whenFalse != "" {
			return stmtFact{kind: factNarrow, nullness: facts.whenFalse}
		}
		// Branch-merge: if (x == null) x = <non-null>;  -> both the then-branch (just
		// assigned) and the fall-through (COND false, non-null) agree x is non-null.
		if ifs.ElseStatement == nil && facts.whenTrue == NullnessNullable &&
			!definitelyExits(ifs.ThenStatement) &&
			branchTrailingNullness(ifs.ThenStatement, symbol, resolve, exprNullness) == NullnessNonNull {
			return stmtFact{kind: factNarrow, nullness: NullnessNonNull}
		}
		// Soundness: any other assignment to x inside a branch leaves its state
		// unprovable here, so earlier facts must not leak past this if.
		if branchAssignsSymbol(ifs.ThenStatement, symbol, resolve) ||
			(ifs.ElseStatement != nil && branchAssignsSymbol(ifs.ElseStatement, symbol, resolve)) {
			return stmtFact{kind: factReset}
		}
	case TryStatement:
		t := stmt.AsTryStatement()
		// Reaching code after a try means the try block completed normally: any thrown
		// exception was caught and the catch did not fall through. So a fact the try body
		// proves at its end (e.g. Objects.requireNonNull(x)) holds afterward - but only
		// when every catch exits and a finally does not reassign the symbol.
		if t.CatchClauses != nil {
			for _, c := range t.CatchClauses.Nodes {
				if !definitelyExits(c.AsCatchClause().Block) {
					return stmtFact{kind: factNone}
				}
			}
		}
		if t.FinallyBlock != nil && branchAssignsSymbol(t.FinallyBlock, symbol, resolve) {
			return stmtFact{kind: factReset}
		}
		stmts := t.TryBlock.AsBlock().Statements.Nodes
		return scanPrecedingFacts(stmts, len(stmts), symbol, resolve, exprNullness)
	}
	return stmtFact{kind: factNone}
}

// assignedValue returns the RHS of `x = <expr>` as a statement, or nil.
func assignedValue(stmt *Node, symbol *Symbol, resolve resolveRef) *Node {
	if stmt.Kind != ExpressionStatement {
		return nil
	}
	expr := stmt.AsExpressionStatement().Expression
	if expr.Kind != AssignmentExpression {
		return nil
	}
	a := expr.AsAssignmentExpression()
	if a.OperatorToken != EqualsToken || !refersToSymbol(a.Left, symbol, resolve) {
		return nil
	}
	return a.Right
}

// branchStatements is the statements of a branch (a block's, or the statement itself).
func branchStatements(branch *Node) []*Node {
	if branch.Kind == Block {
		return branch.AsBlock().Statements.Nodes
	}
	return []*Node{branch}
}

func branchAssignsSymbol(branch *Node, symbol *Symbol, resolve resolveRef) bool {
	for _, s := range branchStatements(branch) {
		if assignedValue(s, symbol, resolve) != nil {
			return true
		}
	}
	return false
}

// branchTrailingNullness is the nullness a branch leaves the symbol with when its
// last statement assigns it, else "" (we only reason about a trailing assignment).
func branchTrailingNullness(branch *Node, symbol *Symbol, resolve resolveRef, exprNullness exprNullnessFn) Nullness {
	stmts := branchStatements(branch)
	if len(stmts) == 0 {
		return ""
	}
	if value := assignedValue(stmts[len(stmts)-1], symbol, resolve); value != nil {
		return exprNullness(value)
	}
	return ""
}

// narrowNullnessAt returns the narrowed nullness of symbol at use site `use`, or ""
// when nothing is proven (the declared nullness then applies). Caller must only pass
// locals/params.
func narrowNullnessAt(use *Node, symbol *Symbol, resolve resolveRef, exprNullness exprNullnessFn) Nullness {
	node := use
	for parent := node.Parent; parent != nil; node, parent = parent, parent.Parent {
		// Preceding statements in an enclosing block (nearest first). An assignment
		// or guard here is checked before the enclosing condition, so a write between
		// a guard and the use correctly invalidates the guard.
		if parent.Kind == Block {
			stmts := parent.AsBlock().Statements
			idx := indexOfNode(stmts, node)
			fact := scanPrecedingFacts(stmts.Nodes, idx, symbol, resolve, exprNullness)
			if fact.kind == factNarrow {
				return fact.nullness
			}
			if fact.kind == factReset {
				return ""
			}
			continue
		}
		// Conditional branch position: inside a branch whose guard proves a fact.
		switch parent.Kind {
		case IfStatement:
			ifs := parent.AsIfStatement()
			f := conditionImplies(ifs.Condition, symbol, resolve)
			if node == ifs.ThenStatement && f.whenTrue != "" {
				return f.whenTrue
			}
			if node == ifs.ElseStatement && f.whenFalse != "" {
				return f.whenFalse
			}
		case ConditionalExpression:
			c := parent.AsConditionalExpression()
			f := conditionImplies(c.Condition, symbol, resolve)
			if node == c.WhenTrue && f.whenTrue != "" {
				return f.whenTrue
			}
			if node == c.WhenFalse && f.whenFalse != "" {
				return f.whenFalse
			}
		case BinaryExpression:
			b := parent.AsBinaryExpression()
			if node == b.Right {
				switch b.OperatorToken {
				case AmpersandAmpersandToken:
					if f := conditionImplies(b.Left, symbol, resolve).whenTrue; f != "" {
						return f
					}
				case BarBarToken:
					if f := conditionImplies(b.Left, symbol, resolve).whenFalse; f != "" {
						return f
					}
				}
			}
		case WhileStatement:
			// The body runs only when the condition held. (A do-while body runs once
			// before the test, so its condition is intentionally not consulted.)
			w := parent.AsWhileStatement()
			if node == w.Statement {
				if f := conditionImplies(w.Condition, symbol, resolve).whenTrue; f != "" {
					return f
				}
			}
		case ForStatement:
			fs := parent.AsForStatement()
			if node == fs.Statement && fs.Condition != nil {
				if f := conditionImplies(fs.Condition, symbol, resolve).whenTrue; f != "" {
					return f
				}
			}
		case SwitchClause:
			// Inside a switch arm: when the switch has a `case null` label, the selector
			// is non-null in every other arm. A reassignment earlier in the arm wins.
			clause := parent.AsSwitchClause()
			var selector *Node
			var clauses *NodeArray
			switch sw := parent.Parent; {
			case sw == nil:
			case sw.Kind == SwitchStatement:
				selector, clauses = sw.AsSwitchStatement().Expression, sw.AsSwitchStatement().Clauses
			case sw.Kind == SwitchExpression:
				selector, clauses = sw.AsSwitchExpression().Expression, sw.AsSwitchExpression().Clauses
			}
			if selector != nil && clause.Statements != nil {
				idx := indexOfNode(clause.Statements, node)
				if idx >= 0 && refersToSymbol(selector, symbol, resolve) {
					fact := scanPrecedingFacts(clause.Statements.Nodes, idx, symbol, resolve, exprNullness)
					if fact.kind == factNarrow {
						return fact.nullness
					}
					if fact.kind == factReset {
						return ""
					}
					if !clauseMatchesNull(clause) && switchHasNullCase(clauses) {
						return NullnessNonNull
					}
				}
			}
		}
	}
	return ""
}

File: src/compiler/narrowing.ts
// Flow-aware nullness narrowing (nikeee/cappu#25). A syntactic dominator walk - not
// a full control-flow graph - that figures out what the code has proven about a
// local/parameter at a use site, the analog of TypeScript's getTypeOfSymbolAtLocation.
// Covers the common Java idioms: if/else guards, early-exit (if (x==null) return),
// && / || short-circuit, ternary, instanceof, Objects.requireNonNull / assert, and
// reassignment. Loops' back-edges and branch-merge of reassignments are not modeled.

import {
  type BinaryExpression,
  type CallExpression,
  type ConditionalExpression,
  type Identifier,
  type IfStatement,
  type InstanceofExpression,
  type Node,
  type ParenthesizedExpression,
  type PrefixUnaryExpression,
  type PropertyAccessExpression,
  type Block,
  type AssertStatement,
  type ExpressionStatement,
  type WhileStatement,
  type ForStatement,
  type Symbol,
  SyntaxKind,
} from "./types.ts";
import { type Nullness } from "./nullness.ts";

/** Resolve a bare reference node to its symbol (the checker's resolveName). */
export type ResolveRef = (node: Node) => Symbol | undefined;
/** The provable nullness of a value expression (null literal, `new`, @NonNull, ...). */
export type ExprNullness = (node: Node) => Nullness | undefined;

/** What a boolean condition proves about `symbol` when it is true / when false. */
interface Facts {
  readonly whenTrue?: Nullness;
  readonly whenFalse?: Nullness;
}

const isNullLiteral = (n: Node): boolean => n.kind === SyntaxKind.NullKeyword;

// A bare reference (`x`) to the tracked symbol. Locals/params are referenced by a
// plain Identifier, so a qualified `this.x` (a field) is intentionally not matched.
function refersToSymbol(node: Node, symbol: Symbol, resolveRef: ResolveRef): boolean {
  return node.kind === SyntaxKind.Identifier && resolveRef(node as Identifier) === symbol;
}

// The simple (unqualified) name of a call's callee, for Objects.requireNonNull etc.
function calleeName(call: CallExpression): string | undefined {
  const callee = call.expression;
  if (callee.kind === SyntaxKind.Identifier) return (callee as Identifier).text;
  if (callee.kind === SyntaxKind.PropertyAccessExpression) {
    return (callee as PropertyAccessExpression).name.text;
  }
  return undefined;
}

function conditionImplies(cond: Node, symbol: Symbol, resolveRef: ResolveRef): Facts {
  switch (cond.kind) {
    case SyntaxKind.ParenthesizedExpression:
      return conditionImplies((cond as ParenthesizedExpression).expression, symbol, resolveRef);
    case SyntaxKind.PrefixUnaryExpression: {
      const u = cond as PrefixUnaryExpression;
      if (u.operator !== SyntaxKind.ExclamationToken) return {};
      const f = conditionImplies(u.operand, symbol, resolveRef);
      return { whenTrue: f.whenFalse, whenFalse: f.whenTrue };
    }
    case SyntaxKind.BinaryExpression: {
      const b = cond as BinaryExpression;
      const op = b.operatorToken;
      if (op === SyntaxKind.EqualsEqualsToken || op === SyntaxKind.ExclamationEqualsToken) {
        const isNullCheck =
          (refersToSymbol(b.left, symbol, resolveRef) && isNullLiteral(b.right)) ||
          (refersToSymbol(b.right, symbol, resolveRef) && isNullLiteral(b.left));
        if (!isNullCheck) return {};
        // `x == null`: true => null, false => non-null. `x != null`: the inverse.
        return op === SyntaxKind.EqualsEqualsToken
          ? { whenTrue: "nullable", whenFalse: "nonNull" }
          : { whenTrue: "nonNull", whenFalse: "nullable" };
      }
      if (op === SyntaxKind.AmpersandAmpersandToken) {
        const l = conditionImplies(b.left, symbol, resolveRef);
        const r = conditionImplies(b.right, symbol, resolveRef);
        return { whenTrue: l.whenTrue ?? r.whenTrue };
      }
      if (op === SyntaxKind.BarBarToken) {
        const l = conditionImplies(b.left, symbol, resolveRef);
        const r = conditionImplies(b.right, symbol, resolveRef);
        return { whenFalse: l.whenFalse ?? r.whenFalse };
      }
      return {};
    }
    case SyntaxKind.InstanceofExpression: {
      const io = cond as InstanceofExpression;
      // `x instanceof T` is false for null, so true proves x non-null.
      return refersToSymbol(io.expression, symbol, resolveRef) ? { whenTrue: "nonNull" } : {};
    }
    case SyntaxKind.CallExpression: {
      const call = cond as CallExpression;
      const arg0 = call.arguments[0];
      if (!arg0 || !refersToSymbol(arg0, symbol, resolveRef)) return {};
      const name = calleeName(call);
      if (name === "nonNull") return { whenTrue: "nonNull" }; // Objects.nonNull(x)
      if (name === "isNull") return { whenTrue: "nullable", whenFalse: "nonNull" }; // Objects.isNull(x)
      return {};
    }
    default:
      return {};
  }
}

// A statement definitely completes abruptly (so code after it is unreachable):
// return/throw/break/continue, or a block whose last statement does.
function definitelyExits(stmt: Node | undefined): boolean {
  if (!stmt) return false;
  switch (stmt.kind) {
    case SyntaxKind.ReturnStatement:
    case SyntaxKind.ThrowStatement:
    case SyntaxKind.BreakStatement:
    case SyntaxKind.ContinueStatement:
      return true;
    case SyntaxKind.Block: {
      const s = (stmt as Block).statements;
      return s.length > 0 && definitelyExits(s[s.length - 1]);
    }
    default:
      return false;
  }
}

// `x = <expr>` as a statement, assigning the tracked symbol.
function assignedValue(stmt: Node, symbol: Symbol, resolveRef: ResolveRef): Node | undefined {
  if (stmt.kind !== SyntaxKind.ExpressionStatement) return undefined;
  const expr = (stmt as ExpressionStatement).expression as Node;
  if (expr.kind !== SyntaxKind.AssignmentExpression) return undefined;
  const a = expr as BinaryExpression;
  if (a.operatorToken !== SyntaxKind.EqualsToken || !refersToSymbol(a.left, symbol, resolveRef)) {
    return undefined;
  }
  return a.right;
}

// The statements of a branch (a block's, or the single statement itself).
function branchStatements(branch: Node): readonly Node[] {
  return branch.kind === SyntaxKind.Block ? (branch as Block).statements : [branch];
}

// Whether a branch contains a top-level assignment to the tracked symbol.
function branchAssignsSymbol(branch: Node, symbol: Symbol, resolveRef: ResolveRef): boolean {
  return branchStatements(branch).some(s => assignedValue(s, symbol, resolveRef));
}

// The nullness a branch leaves the symbol with when its last statement assigns it,
// else undefined (we only reason about a trailing assignment).
function branchTrailingNullness(
  branch: Node,
  symbol: Symbol,
  resolveRef: ResolveRef,
  exprNullness: ExprNullness,
): Nullness | undefined {
  const stmts = branchStatements(branch);
  const last = stmts[stmts.length - 1];
  const value = last && assignedValue(last, symbol, resolveRef);
  return value ? exprNullness(value) : undefined;
}

// A `return`/`undefined` from one preceding statement: "narrow" gives a proven
// nullness; "reset" means an assignment whose value we could not prove (the symbol's
// state is unknown from here, so earlier facts no longer apply); undefined = keep looking.
type StmtFact = { kind: "narrow"; nullness: Nullness } | { kind: "reset" } | undefined;

function precedingStatementFact(
  stmt: Node,
  symbol: Symbol,
  resolveRef: ResolveRef,
  exprNullness: ExprNullness,
): StmtFact {
  // x = <expr>;  -> the value's nullness; an assignment supersedes earlier facts.
  if (stmt.kind === SyntaxKind.ExpressionStatement) {
    const expr = (stmt as ExpressionStatement).expression as Node;
    if (expr.kind === SyntaxKind.AssignmentExpression) {
      const a = expr as BinaryExpression;
      if (
        a.operatorToken === SyntaxKind.EqualsToken &&
        refersToSymbol(a.left, symbol, resolveRef)
      ) {
        const n = exprNullness(a.right);
        return n ? { kind: "narrow", nullness: n } : { kind: "reset" };
      }
    }
    // Objects.requireNonNull(x[, ...]);  -> x is non-null afterwards.
    if (expr.kind === SyntaxKind.CallExpression) {
      const call = expr as CallExpression;
      const arg0 = call.arguments[0];
      if (
        calleeName(call) === "requireNonNull" &&
        arg0 &&
        refersToSymbol(arg0, symbol, resolveRef)
      ) {
        return { kind: "narrow", nullness: "nonNull" };
      }
    }
    return undefined;
  }
  // assert x != null;  -> x is non-null afterwards.
  if (stmt.kind === SyntaxKind.AssertStatement) {
    const f = conditionImplies((stmt as AssertStatement).condition, symbol, resolveRef).whenTrue;
    return f === "nonNull" ? { kind: "narrow", nullness: "nonNull" } : undefined;
  }
  if (stmt.kind === SyntaxKind.IfStatement) {
    const ifs = stmt as IfStatement;
    const facts = conditionImplies(ifs.condition, symbol, resolveRef);
    // Early-exit: if (COND) <abrupt>;  -> after the if, only the fall-through path
    // remains, so x has whatever COND-false proves (e.g. if (x==null) return; -> non-null).
    if (!ifs.elseStatement && definitelyExits(ifs.thenStatement) && facts.whenFalse) {
      return { kind: "narrow", nullness: facts.whenFalse };
    }
    // Branch-merge: if (x == null) x = <non-null>;  -> both the then-branch (just
    // assigned) and the fall-through (COND was false, so non-null) agree x is non-null.
    if (
      !ifs.elseStatement &&
      facts.whenTrue === "nullable" &&
      !definitelyExits(ifs.thenStatement) &&
      branchTrailingNullness(ifs.thenStatement, symbol, resolveRef, exprNullness) === "nonNull"
    ) {
      return { kind: "narrow", nullness: "nonNull" };
    }
    // Soundness: any other assignment to x inside a branch leaves its state
    // unprovable here, so earlier facts must not leak past this if.
    if (
      branchAssignsSymbol(ifs.thenStatement, symbol, resolveRef) ||
      (ifs.elseStatement && branchAssignsSymbol(ifs.elseStatement, symbol, resolveRef))
    ) {
      return { kind: "reset" };
    }
  }
  return undefined;
}

/**
 * The narrowed nullness of `symbol` at use site `use`, or undefined when nothing is
 * proven (the declared nullness then applies). Caller must only pass locals/params.
 */
export function narrowNullnessAt(
  use: Node,
  symbol: Symbol,
  resolveRef: ResolveRef,
  exprNullness: ExprNullness,
): Nullness | undefined {
  let node: Node = use;
  for (let parent = node.parent; parent; node = parent, parent = parent.parent) {
    // Preceding statements in an enclosing block (nearest first). An assignment or
    // guard here is checked before the enclosing condition, so a write between a
    // guard and the use correctly invalidates the guard.
    if (parent.kind === SyntaxKind.Block) {
      const stmts = (parent as Block).statements;
      const idx = stmts.indexOf(node as never);
      for (let i = idx - 1; i >= 0; i--) {
        const fact = precedingStatementFact(stmts[i]!, symbol, resolveRef, exprNullness);
        if (fact?.kind === "narrow") return fact.nullness;
        if (fact?.kind === "reset") return undefined;
      }
      continue;
    }
    // Conditional branch position: we are inside a branch whose guard proves a fact.
    if (parent.kind === SyntaxKind.IfStatement) {
      const ifs = parent as IfStatement;
      const facts = conditionImplies(ifs.condition, symbol, resolveRef);
      if (node === ifs.thenStatement && facts.whenTrue) return facts.whenTrue;
      if (node === ifs.elseStatement && facts.whenFalse) return facts.whenFalse;
    } else if (parent.kind === SyntaxKind.ConditionalExpression) {
      const c = parent as ConditionalExpression;
      const facts = conditionImplies(c.condition, symbol, resolveRef);
      if (node === c.whenTrue && facts.whenTrue) return facts.whenTrue;
      if (node === c.whenFalse && facts.whenFalse) return facts.whenFalse;
    } else if (parent.kind === SyntaxKind.BinaryExpression) {
      const b = parent as BinaryExpression;
      if (node === b.right) {
        if (b.operatorToken === SyntaxKind.AmpersandAmpersandToken) {
          const f = conditionImplies(b.left, symbol, resolveRef).whenTrue;
          if (f) return f;
        } else if (b.operatorToken === SyntaxKind.BarBarToken) {
          const f = conditionImplies(b.left, symbol, resolveRef).whenFalse;
          if (f) return f;
        }
      }
    } else if (parent.kind === SyntaxKind.WhileStatement) {
      // The body runs only when the condition held. (A do-while body runs once
      // before the test, so its condition is intentionally not consulted.)
      const w = parent as WhileStatement;
      if (node === w.statement) {
        const f = conditionImplies(w.condition, symbol, resolveRef).whenTrue;
        if (f) return f;
      }
    } else if (parent.kind === SyntaxKind.ForStatement) {
      const fs = parent as ForStatement;
      if (node === fs.statement && fs.condition) {
        const f = conditionImplies(fs.condition, symbol, resolveRef).whenTrue;
        if (f) return f;
      }
    }
  }
  return undefined;
}
